src/Apps/Legion/Components/GraphVisualizer/Hooks/CustomClickHandler/CustomClickHandler.tsx

```diff
@@ -0,0 +1,126 @@
+            // focusNodeById(nodeData.id);
+
+            // set selected node
+            setSelectedNode(nodeData);
+        },
+        [setSelectedNode]
+    );
+
+    // EDGE CALLBACKS
+    // const setSelectedEdge = useCallback(
+    //     (edgeData: ICustomEdgeData) => {
+    //         graphDispatch({
+    //             type: OatPageContextActionType.SET_OAT_SELECTED_MODEL,
+    //             payload: {
+    //                 selection: getSelectionFromGraphItem(edgeData)
+    //             }
+    //         });
+    //     },
+    //     [oatPageDispatch]
+    // );
+    // master handler
+    // const onEdgeSelected = useCallback(
+    //     (nodeData: ICustomEdgeData) => {
+    //         // set selected node
+    //         setSelectedEdge(nodeData);
+    //     },
+    //     [setSelectedEdge]
+    // );
+
+    // side effects
+    // register for node click events
+    useEffect(() => {
+        logDebugConsole('debug', 'Register NODE click handler');
+        const handleClick = (evt: IG6GraphEvent) => {
+            const node = evt.item as INode;
+            const model = node.getModel() as ICustomNodeConfig;
+            logDebugConsole('info', 'Node clicked {node}', model);
+            onNodeSelected(model.data);
+        };
+        graph.on('node:click', handleClick);
+        return () => {
+            logDebugConsole('debug', 'Unregister NODE click handler');
+            graph.off('node:click', handleClick);
+        };
+    }, [graph, onNodeSelected]);
+
+    // register for canvas click events
+    useEffect(() => {
+        const handleCanvasClick = (ev: IG6GraphEvent) => {
+            // logDebugConsole('info', 'Click event {event}', ev);
+            const shape = ev.item;
+            if (!shape || shape.get('type') !== 'node') {
+                logDebugConsole('info', 'Canvas clicked {event}', ev);
+                onNodeSelected(null);
+            }
+        };
+
+        logDebugConsole('debug', 'Register CANVAS click handler');
+        graph.on('click', handleCanvasClick);
+        return () => {
+            logDebugConsole('debug', 'Unregister CANVAS click handler');
+            graph.off('click', handleCanvasClick);
+        };
+    }, [graph, onNodeSelected]);
+
+    // register for edge click events
+    // useEffect(() => {
+    //     logDebugConsole('debug', 'Register EDGE click handler');
+    //     const handleClick = (evt: IG6GraphEvent) => {
+    //         const edge = evt.item as IEdge;
+    //         const model = edge.getModel() as ICustomEdgeConfig;
+    //         logDebugConsole('info', 'Edge clicked {edge}', model);
+    //         onEdgeSelected(model.data);
+    //     };
+
+    //     graph.on('edge:click', handleClick);
+    //     return () => {
+    //         logDebugConsole('debug', 'Unregister EDGE click handler');
+    //         graph.off('edge:click', handleClick);
+    //     };
+    // }, [apis, graph, onEdgeSelected]);
+
+    return null;
```

Comment:
how is CustomClickHandler a "hook"? it returns null and implemented as a component?

Reply:
This is a pattern that the graph library uses. It is not so much a "hook" as it is a way to register event handlers. I couldn't think of a better name. You don't' use these like a traditional hook since they don't produce or return an output, they simply bind to the events from the graph. The reason they include them as a child DOM element of the graph is so that you can be sure they will always have the graph context provided by the parent component.